src/app/data-planes/data/index.ts

```diff
@@ -0,0 +1,202 @@
+import { DataplaneNetworking } from './networking'
+import { Version } from './version'
+import type {
+  PaginatedApiListResponse as CollectionResponse,
+} from '@/types/api.d'
+import type {
+  DataPlaneOverview as OriginalDataplaneOverview,
+  DataPlane as OriginalDataplane,
+  DataPlaneInsight as OriginalDataplaneInsight,
+  DiscoverySubscription,
+  KumaDpVersion,
+  EnvoyVersion,
+} from '@/types/index.d'
+import {
+  compatibilityKind,
+  COMPATIBLE,
+} from '@/utilities/dataplane'
+
+export type Dataplane = OriginalDataplane & {
+  networking: DataplaneNetworking
+  tags: { label: string, value: string }[]
+  service: string
+  status: string
+  protocol: string
+  zone: string
+}
+export type DataplaneInsight = OriginalDataplaneInsight & {
+  subscriptions: DiscoverySubscription[];
+}
+
+export type DataplaneOverview = OriginalDataplaneOverview & {
+  status: string
+  // dependencies: Record<string, string>
+  lastUpdated: Date
+  version: string
+  compatible: boolean,
+  dataplane: Dataplane
+  dataplaneInsight: DataplaneInsight
+}
+
+type Label = {
+  label: string
+  value: string
+}
+type BuildVersion = KumaDpVersion | EnvoyVersion | { version: string}
+///
+
+export const Dataplane = {
+  fromJSONCollection(collection: CollectionResponse<unknown>): CollectionResponse<Dataplane> {
```

Comment:
🤔  let me look at this again its been a good few months since I originally wrote this. I think its possibly to do with the fact that I want to pass in completely untyped data and this does the reshaping and typing.

Reply:
Yeah so, eventually completely untyped data will come into these straight from `fetch` (or other HTTP retrieval methods) and that will be fed straight into `fromJSONCollection` or `fromJSON`. _These will in turn call `fromJSON` on the generated types first_ (we don't do this currently, but we do cast the type), before doing our "frontend type amends" and then finally returning the result back to the application all reshaped and typed correctly for our frontend.

Let me know if you have any other questions here and I can explain a little more if necessary.

Reply:
I think that’s a problem.

What you’re proposing is the following "data pipeline":

```
fetch
  ----> data with `unknown` type
  --(apply fromJSON* transforms)--> data with normalized "UI-ready" type
```

I think it should work like this:

```
fetch
  ----> data with (ideally generated) API type
  --(apply fromJSON* transforms)--> data with normalized "UI-ready" type
```

I.e. fetch data is assigned an (ideally generated) API type and we transform based on _that_ type.

This would be necessary to spot and incorporate API changes.

Reply:
What I will be proposing is the second thing you have said, just organised slightly differently.

```
fetch
  --(call fromJSON* transforms)--> which firstly
    -- (apply fromJSON* transforms from the generated types) --> and then
    -- (apply fromJSON* transforms)--> data with normalized "UI-ready" type --> returns the correctly typed, correctly shaped data to the view via DataSource
```

Reply:
Just wanted to note that for a more guarantees of type safety application, it is essential that we have generated types. If we continue to hand write them we are just moving the bugs to a different layer i.e. Typescript.

Reply:
To make sure I’m following your example, you seem to envision two steps:

1. from fetch response body of type `unknown` to API type.
2. from value with API type to value with UI type.

Is that correct?

If so:

- In which step does normalization of things like omitempty array fields happen? Step 2, yes?
- Which of these steps matches what you intend `fromJSON*` to do?

Reply:
> from fetch response body of type unknown to API type.

To be exact it could be `any` but I'd understood that `unknown` would be better to use than `any` but I'm 🤷 

But yes, `fetch -> generated shape/type -> frontend amended shape/type -> datasource response`

> In which step does normalization of things like omitempty array fields happen? Step 2, yes?

Anything that has `undefined` or `null`  in the generated shape/type that would potentially cause us problems on the frontend would be done in step 2 yes. For example converting an `string[] | undefined` to just `string[]` (defaulting to `[]`)

> Which of these steps matches what you intend fromJSON* to do?

Both. `Dataplane.fromJSON` firstly calls `GeneratedDataplane.fromJSON` and then on the generated shape/type of the result of that, then performs additional frontend shaping/typing, then the return of `Dataplane.fromJSON` is the result of that.

Reply:
> In our context, it makes little difference. We will always have a certain point where we make a type assertion that says "a fetch call to this known API path yields a structure of that type".

Yes exactly, but I wanted to be absolutely clear.

---

> Then it looks like the fromJSON* input is inadvertently the generated API type and should be what we base the transformations on, not unknown

Its starts off as `unknown` gets converted into the generated type, and only then gets amended.

We are both saying exactly the same thing here.

Reply:
> I really think it’s critical for this approach to make sense that we rely on type-checking from API type to UI type and implement doing so in the fromJSON* methods.

Yes exactly, its important to rely on the generated types instead of just going from unknown to our own frontend hand written types.

Reply:
> These will in turn call fromJSON on the generated types first (we don't do this currently, but we do cast the type), before doing our "frontend type amends" and then finally returning the result back to the application all reshaped and typed correctly for our frontend.

https://github.com/kumahq/kuma-gui/pull/1242#discussion_r1276028309

Reply:
Maybe this helps:

```javascript
Dataplane.fromJSON = (obj: unknown) => {
  const dataplane = GeneratedDataplane.fromJSON()
  // fix up dataplane with frontend specific things here
  dataplane.frontendUsefulproperty = reshapedTheThing(dataplane)
  dataplane.oopsie = !Array.isArray(dataplane.oopsie) ? [] : dataplane.oopsie
  return dataplane;
}
```

Reply:
Then I guess I don’t understand why we’re opting out of the type system here.

The objects passed into the `fromJSON*` methods should have type annotations for the API types (i.e. `Original*` types), not `unknown`. From this, we can do data transformations like we do today.

The code in the transformers here instead uses explicit type assertions at the start:

```ts
     const item = {
       ...(object as OriginalDataplaneOverview),
     } as DataplaneOverview
```

Following this code, you no longer access `object` and only `item`. You told the TypeScript compiler that `item` is already of the new `DataplaneOverview` type (i.e. a UI type).

As a consequence, you can delete *ALL* of the function’s body after the assignment of `item` up until the final `return` and no errors pop up:

```ts
    const item = {
      ...(object as OriginalDataplaneOverview),
    } as DataplaneOverview
    return item
```

Effectively, the current approach removes all the type safety we have in the `transformToTableData` functions.

Reply:
Oh well thats a mistake then, thankyou for spotting!

Reply:
Hmm yeah this is the stuff from the other place using the generated types (`DataplaneOverviewProto`)

<img width="790" alt="Screenshot 2023-07-27 at 13 28 08" src="https://github.com/kumahq/kuma-gui/assets/554604/50b3759e-c200-4b3c-b5af-997d8db5c0d9">

I think I might have rewired things when I've moved things over here without the generated types. Let me take look

Reply:
Hmm I actually think if you are going to delete the data reshaping code you should probably delete the type also. Feels a little weird to me, but anyway. I've pushed a commit which think addresses your concerns without bringing in all the generated types. Let me know of thats better.

Reply:
This is slightly better, but there are still a lot of `unknown`s in here that can be avoided and doing so brings type safety closer to what we had before.

Changing this code so that the `fromJSON*` methods take objects typed with their Original* type and avoiding type assertions to those types in the function bodies reveals errors hidden by those type assertions.

For example, in `DataplaneOverview.fromJSON`, the call `Dataplane.fromJSON(object.dataplane)` makes the incorrect assumption that `object.dataplane` is of type `OriginalDataplane`. While that seems reasonable to expect, it’s not true: `OriginalDataplaneOverview['dataplane']` is typed as an object with only a `networking` field while `OriginalDataplane` represents the complete `Dataplane` resource. I verified that this is true with the Kuma API.

I’m putting the relatively small changes I made to your code in my local copy to show these errors below. I think it would be much more robust if the `fromJSON*` methods expect their corresponding `Original*` type.

<details>
<summary>Changes</summary>

```ts
export const Dataplane = {
  fromJSONCollection(collection: CollectionResponse<OriginalDataplane>): CollectionResponse<Dataplane> {
    collection.items = collection.items.map((item) => {
      return Dataplane.fromJSON(item)
    })
    return collection as CollectionResponse<Dataplane>
  },
  fromJSON(object: OriginalDataplane): Dataplane {
    const item: Dataplane = {
      ...object,
      // make sure networking.inbound is at least an empty array
      networking: DataplaneNetworking.fromJSON(object.networking),
      //
      tags: [],
      service: '',
      protocol: '',
      zone: '',
    }
    item.tags = getTags(item)
    item.service = getTag(item, 'kuma.io/service')
    item.protocol = getTag(item, 'kuma.io/protocol')
    item.zone = getTag(item, 'kuma.io/zone')
    return item
  },
}

export const DataplaneInsight = {
  fromJSON(object: OriginalDataplaneInsight): DataplaneInsight {
    const item: DataplaneInsight = {
      ...object,
      // make sure subscriptions is at least an empty array
      subscriptions: Array.isArray(object?.subscriptions) ? object.subscriptions : [],
    }
    return item
  },
}
export const DataplaneOverview = {
  fromJSONCollection(collection: CollectionResponse<OriginalDataplaneOverview>): CollectionResponse<DataplaneOverview> {
    collection.items = collection.items.map((item) => {
      return DataplaneOverview.fromJSON(item)
    })
    return collection as CollectionResponse<DataplaneOverview>
  },
  fromJSON(object: OriginalDataplaneOverview): DataplaneOverview {
    const item: DataplaneOverview = {
      ...object,
      dataplane: Dataplane.fromJSON(object.dataplane),
      dataplaneInsight: DataplaneInsight.fromJSON(object.dataplaneInsight || {}),
      status: '',
      lastUpdated: new Date(0),
      version: '',
      compatible: false,
    }

    const dependencies = getDependencies(item.dataplaneInsight)
    item.lastUpdated = getLastUpdated(item.dataplaneInsight)
    item.version = dependencies.kumaDp.version
    const compatibility = compatibilityKind(
      {
        kumaDp: dependencies.kumaDp as KumaDpVersion,
        envoy: dependencies.envoy as EnvoyVersion,
        dependencies: {} as Record<string, string>,
      },
    ).kind
    item.compatible = compatibility !== COMPATIBLE
    item.status = getStatus(item)
    return item
  },
}
```

</details>

Reply:
Hmm, ok this is giving me type errors? Let me look again, maybe I've done something wrong.

In the meantime, I think a larger point is, when we remove `kumaAPI` and replace it with a HTTP client that HTTP client will return `any/unknown`:


```javascript
export const sources = (request: HTTPClient) => {
  return {
    '/dataplanes': async () => {
      return DataplaneOverview.fromJSONCollection(
        await request({...})
      }
    },
}
```

and all data manipulation will live in `fromJSON`, so we don't have some data manipulation in the thing that does the requesting and some in the thing that does the majority of the data manipulation.

i.e. These `fromJSON` method should receive `any/unknown`.

Reply:
> Hmm, ok this is giving me type errors?

Yes, it does. It’s the expected errors stemming from the fact that `object.dataplane` isn’t actually of type `DataplaneOverview`.

> i.e. These fromJSON method should receive any/unknown.

Well, not quite. `fromJSON*` should still receive the Original* types, but the caller site will pass in `any`/`unknown` data (i.e. a `fetch` response body). Basically, the function signature of `fromJSON*` takes the place of the return type annotation of our current KumaApi methods. Those are effectively type assertions from `any` to the corresponding Original* type.

Reply:
@kleinfreund Can I check if there is anything else in this PR that needs addressing apart from this?

Reply:
I went over this PR again to make sure I added comments for everything I wanted to address now.

Reply:
Ok just to be clear, you said "everything you wanted to address _now_". Was that "now" as in you have other things for later, or "now" as in, "I hadn't addressed everything before and now I have"?

Reply:
@kleinfreund I'm assuming that this is the last unresolved conversation. Could you take a look at the last commit here, which moves all data reshaping/retyping to the same place inside the toJSON/toJSONCollection methods of the data layer.

https://github.com/kumahq/kuma-gui/pull/1242/commits/4a2dc3b82ccde437862be78c40edb43e9fd0d217

It has the additional benefit of us being able to work towards removing the entire KumaAPI class and moving to individual HTTP modules (one per data type), which I believe is desirable. I actually wanted to start doing this at a later date, but I think it might help to resolve this conversation.

Lmk

Reply:
The issues I’ve outlined in https://github.com/kumahq/kuma-gui/pull/1242#discussion_r1277255689 still exists. The use of type assertions in the function body combined with the function parameter being of type `unknown` hide away important type information in the transition from API type to UI type.

While this might not immediately result in application errors, it makes it easier to introduce them. Code in `Dataplane.fromJSON` will, by its own type assertion, assume `object` to have the type `OriginalDataplane`, but as I pointed out, this **will not be true** when provided with `dataplaneOverview.dataplane` which only has a `networking` property.

Reply:
Why can't we have types in the signature?

Reply:
I made a few more amends, as I didn't want to bring in the generated types here I had to add them myself and I made an oopise and left off `insights` I think?

> While this might not immediately result in application errors, it makes it easier to introduce them. Code in Dataplane.fromJSON will, by its own type assertion, assume object to have the type OriginalDataplane, but as I pointed out, this will not be true when provided with dataplaneOverview.dataplane which only has a networking property.

I think what we are missing are the generated type converters which would go in in place of `(object as OriginalDataplane)` and instead be `OriginalDataplane.fromJSON(object)`. These generated type converters will also accept completely empty objects and converter them into a valid object/type for you like `OriginalDataplane.fromJSON({})`. Ideally I was hoping to add all that in separately and then we can delete all the handwritten ones and depend on the generated ones plus the ui specific types.

lmk

Reply:
> Why can't we have types in the signature?

Because they are coming straight from `fetch` and because of that the type is `unknown/any`.

Reply:
...and maybe related. You need to be able to say `Dataplane.fromJSON({})` to get a valid dataplane object.

FYI, I'm just copying what the generated types/reshaping does here.

Reply:
> Because the data is coming straight from `fetch` and due to that the type is `unknown/any`.

True. Then the course of action is to call the appropriate `fromJSON*` method either without any changes (which already works as the type of the responde body will be `any`) or with the appropriate type assertion (if we change our fetch to return `unknown`).

In short: You can already define the correct types in the signature of the `fromJSON*` methods and work type-safely from API type to UI type.

I’ve checked out your latest changes and the issues I mentioned in https://github.com/kumahq/kuma-gui/pull/1242#discussion_r1277255689 and https://github.com/kumahq/kuma-gui/pull/1242#discussion_r1277530249 still exist.

> You need to be able to say Dataplane.fromJSON({}) to get a valid dataplane object (super handy for testing)

I strongly oppose that this is a use case we should aim at supporting currently. Getting an empty object in a `fromJSON*` method *when the associated API type doesn’t allow this* is in fact something that should immediately lead to errors. This is not a application scenario we’re currently facing and so we shouldn’t optimize for it.

If you want to be able to use this mechanism for easier testing, this logic must remain outside the application code path.

Reply:
> Yeah I would have expected the type casting (response as OriginalDataplane) to happen in the fetch thus you'd have a type here.

I suppose it kinda does, the other option would be to add the type cast here:

https://github.com/kumahq/kuma-gui/pull/1242/files#diff-8c51094d188779df11ec5a418503cf5a77c088bb12581335955952e7deef4c3dR33

So:

```javascript
const res = (await request.get(`/meshes/${params.mesh}/dataplanes+insights`) as Dataplane)
```

But again, I think that will cause issues in the future with the generated types as the generated ones I have all expect `any` to be passed through to `fromJSON`.

> Dataplane.fromJSON({}) is funky if it's only for testing does it make sense to be mixed in regular code?

Maybe not, we can make the generated property/type checkers private in someway so you can't get to them from the application itself. This functionality is generated functionality, I'm just copying it over for now.

Reply:
> I suppose it kinda does, the other option would be to add the type cast here:

This has the benefit that it would allow us to change `RestClient` to return response bodies as `unknown`. That being said, for now, `any` is totally fine.

Reply:
I'd much rather go with the type casting above.

When you are talking about type generated. Generated from what? In any case these are easy things to fix later so let's do what makes most sense now (having methods with correct types) and we can always pedal back if this is too limiting in the future no?

Reply:
Ok sounds good, lemme look at doing that.

To generate the types from kuma I was using `ts-proto` along with something like:

```make
protots/%:
	mkdir -p ./generated/pb/
	cd $(KUMA_DIR) && \
	$(PROTOC) \
	  --plugin=../kuma-utils/packages/kuma-api-types/node_modules/.bin/protoc-gen-ts_proto \
    --ts_proto_opt=esModuleInterop=true \
    --ts_proto_opt=useOptionals=none \
    --ts_proto_opt=initializeFieldsAsUndefined=false \
    --ts_proto_opt=outputEncodeMethods=false \
    --ts_proto_opt=outputJSONMethods=true \
    --ts_proto_opt=useExactTypes=false \
    --ts_proto_opt=usePrototypeForDefaults=true \
    --ts_proto_opt=stringEnums=true \
    --ts_proto_opt=outputClientImpl=false \
    --ts_proto_opt=nestJS=false \
    --ts_proto_opt=browser \
    --ts_proto_out=../kuma-utils/packages/kuma-api-types/generated/pb/ \
		api/$*/*.proto

```

Reply:
Ok so if you go with generated types in the future we'll use open api specs as we're slowly moving away from protos. So let's no get too attached to what this generator does.

Reply:
Ah yeah good point, fyi this is half the reason why would prefer all this in its own layer.

Lemme change this up brb

Reply:
😬  I'm hitting issues now because a lot of the "sub converters" won't take `Thing | undefined`. I think I can maybe change everything so that the undefined removal is done outside of the "sub converters" but that means that datashaping of say `DataplaneInsight` will be done outside of the `DataplaneInsight.fromJSON()` function, which isn't ideal.

I think either way I don't think I'm going to have time to do this now.

Reply:
For reference, I wanted to provide my rough version of what I would change to get rid of the `unknown` types in the data transformation code. It doesn’t cover everything in this PR, but it does away with the type errors in `src/app/data-planes/data/index.ts`.

(Paste over the respective files for the diff)

`src/app/data-planes/data/index.ts`:

<details>
<summary>Code</summary>

```ts
import { DataplaneNetworking } from './networking'
import { Version } from './version'
import type {
  PaginatedApiListResponse as CollectionResponse,
} from '@/types/api.d'
import type {
  DataPlaneOverview as OriginalDataplaneOverview,
  DataPlane as OriginalDataplane,
  DataPlaneInsight as OriginalDataplaneInsight,
  DiscoverySubscription,
  KumaDpVersion,
  EnvoyVersion,
} from '@/types/index.d'
import {
  compatibilityKind,
  COMPATIBLE,
} from '@/utilities/dataplane'

export type Dataplane = OriginalDataplane & {
  networking: DataplaneNetworking
  tags: { label: string, value: string }[]
  service: string
  protocol: string
  zone: string
}
export type DataplaneInsight = OriginalDataplaneInsight & {
  subscriptions: DiscoverySubscription[];
}

export type DataplaneOverview = OriginalDataplaneOverview & {
  status: string
  // dependencies: Record<string, string>
  lastUpdated: Date
  version: string
  compatible: boolean,
  dataplane: {
    networking: DataplaneNetworking
  }
  dataplaneInsight: {
    subscriptions: DiscoverySubscription[];
  }
}

type Label = {
  label: string
  value: string
}
type BuildVersion = KumaDpVersion | EnvoyVersion | { version: string}
///

export const Dataplane = {
  fromJSONCollection(collection: CollectionResponse<OriginalDataplane>): CollectionResponse<Dataplane> {
    collection.items = collection.items.map((item) => {
      return Dataplane.fromJSON(item)
    })
    return collection as CollectionResponse<Dataplane>
  },
  fromJSON(object: OriginalDataplane): Dataplane {
    const item: Dataplane = {
      ...object,
      // make sure networking.inbound is at least an empty array
      networking: DataplaneNetworking.fromJSON(object.networking),
      //
      tags: [],
      service: '',
      protocol: '',
      zone: '',
    }
    item.tags = getTags(item)
    item.service = getTag(item, 'kuma.io/service')
    item.protocol = getTag(item, 'kuma.io/protocol')
    item.zone = getTag(item, 'kuma.io/zone')
    return item
  },
}

export const DataplaneInsight = {
  fromJSON(object: OriginalDataplaneInsight | undefined): DataplaneInsight {
    const item: DataplaneInsight = {
      ...object,
      // make sure subscriptions is at least an empty array
      subscriptions: object && Array.isArray(object?.subscriptions) ? object.subscriptions : [],
    }
    return item
  },
}
export const DataplaneOverview = {
  fromJSONCollection(collection: CollectionResponse<OriginalDataplaneOverview>): CollectionResponse<DataplaneOverview> {
    collection.items = collection.items.map((item) => {
      return DataplaneOverview.fromJSON(item)
    })
    return collection as CollectionResponse<DataplaneOverview>
  },
  fromJSON(object: OriginalDataplaneOverview): DataplaneOverview {
    const item: DataplaneOverview = {
      ...object,
      dataplane: {
        networking: DataplaneNetworking.fromJSON(object.dataplane.networking),
      },
      dataplaneInsight: DataplaneInsight.fromJSON(object.dataplaneInsight),
      status: '',
      lastUpdated: new Date(0),
      version: '',
      compatible: false,
    }

    const dependencies = getDependencies(item.dataplaneInsight)
    item.lastUpdated = getLastUpdated(item.dataplaneInsight)
    item.version = dependencies.kumaDp.version
    const compatibility = compatibilityKind(
      {
        kumaDp: dependencies.kumaDp as KumaDpVersion,
        envoy: dependencies.envoy as EnvoyVersion,
        dependencies: {} as Record<string, string>,
      },
    ).kind
    item.compatible = compatibility !== COMPATIBLE
    item.status = getStatus(item)
    return item
  },
}

// utilities
function getStatus(data: DataplaneOverview): string {
  const dataplane = data.dataplane
  const dataplaneInsight = data.dataplaneInsight
  const errors = dataplane.networking.inbound
    .filter(item => {
      return item.health && !item.health.ready
    })
    .map((item: {port: any; tags: {[x: string]: any;};}) => `Inbound on port ${item.port} is not ready (kuma.io/service: ${item.tags['kuma.io/service']})`)

  let status: string
  switch (true) {
    case dataplane.networking.inbound.length === 0:
      status = 'online'
      break
    // if errors and inbounds are equal, even if they are both 0
    // then we are offline
    case errors.length === dataplane.networking.inbound.length:
      status = 'offline'
      break
    // otherwise any errors at all, we are degraded
    case errors.length > 0:
      status = 'partially_degraded'
      break
    default: {
      const proxyOnline = dataplaneInsight.subscriptions.some(
        (item: DiscoverySubscription) => item.connectTime && item.connectTime.length && !item.disconnectTime,
      )
      status = proxyOnline ? 'online' : 'offline'
    }
  }
  return status
}
function getTag(dataplane: Dataplane, key: string): string {
  return getTags(dataplane).find((item: { label: string }) => item.label === key)?.value ?? ''
}
function getTags(dataplane: Dataplane): Label[] {
  let tags: [string, string][] = []
  switch (true) {
    case dataplane.networking.inbound.length > 0:
      tags = dataplane.networking.inbound
        .filter((inbound) => 'tags' in inbound)
        .flatMap((inbound) => Object.entries(inbound.tags))
      break
    case Object.keys(dataplane.networking.gateway?.tags || {}).length > 0:
      // gateway data plane has no inbounds, but has tags embedded in gateway branch
      tags = Object.entries(dataplane.networking.gateway?.tags ?? {})
      break
  }

  const uniqueTags = Array.from(new Set(tags.map(([key, value]) => `${key}=${value}`)))
  uniqueTags.sort((tagPairA: string, tagPairB: string) => tagPairA.localeCompare(tagPairB))
  return uniqueTags
    .map((tagPair: string) => tagPair.split('='))
    .map(([label, value]) => ({ label, value }))
}
function getVersion(item: DataplaneInsight): Version {
  return item.subscriptions.length === 0 ? Version.fromJSON({}) : Version.fromJSON(item.subscriptions[item.subscriptions.length - 1].version)
}

function getDependencies(item: DataplaneInsight): Record<string, BuildVersion> {
  const version = getVersion(item)
  const deps: Record<string, BuildVersion> = {}
  if (version.envoy) {
    deps.envoy = version.envoy
  }
  if (version.kumaDp) {
    deps.kumaDp = version.kumaDp
  }
  if (version.dependencies) {
    Object.entries(version.dependencies).forEach(([key, value]) => {
      deps[key] = {
        version: value,
      }
    })
  }
  return deps
}

function getLastUpdated(insight: DataplaneInsight): Date {
  return new Date(Math.max(
    ...insight.subscriptions.reduce((prev: number[], item: DiscoverySubscription) => {
      return prev.concat([new Date(item.status?.lastUpdateTime)?.getTime() ?? 0])
    }, []),
  ))
}
```

</details>

`src/app/data-planes/data/networking/index.ts`:

<details>
<summary>Code</summary>

```ts
import type {
  DataplaneNetworking as OriginalDataplaneNetworking,
  DataplaneNetworkingInbound as OriginalDataplaneNetworkingInbound,
  DataplaneNetworkingInboundHealth as OriginalDataplaneNetworkingInboundHealth,
} from './original'
export type DataplaneNetworkingInboundHealth = OriginalDataplaneNetworkingInboundHealth & {

}
export type DataplaneNetworkingInbound = OriginalDataplaneNetworkingInbound & {
  health: DataplaneNetworkingInboundHealth
}
export type DataplaneNetworking = OriginalDataplaneNetworking & {
  inbound: DataplaneNetworkingInbound[]
}
export const DataplaneNetworking = {
  fromJSON(object: OriginalDataplaneNetworking): DataplaneNetworking {
    return {
      ...object,
      ...{
        inbound: Array.isArray(object.inbound) ? object.inbound.map((item: any) => DataplaneNetworkingInbound.fromJSON(item || {})) : [],
      },
    }
  },
}
export const DataplaneNetworkingInbound = {
  fromJSON(object: any): DataplaneNetworkingInbound {
    return {
      ...object,
      // an undefined health means we don't know either way
      // health: DataplaneNetworkingInboundHealth.fromJSON(object.health || {}),
    }
  },
}
export const DataplaneNetworkingInboundHealth = {
  fromJSON(object: any): DataplaneNetworkingInboundHealth {
    return { ready: isSet(object.ready) ? Boolean(object.ready) : false }
  },
}
function isSet(value: any): boolean {
  return value !== null && value !== undefined
}
```

</details>

Reply:
I believe this conversation has been addressed, could you let me know what the latest thoughts are? Thanks!

Reply:
> One [...} Consider that part of the conversation resolved.

Awesome, glad we've cleared up that misunderstanding.

> Second: I want to reiterate how making the type signature for the fromJSON* methods accept any (or unknown) is not acceptable.

Ok I not sure I understand your thinking here then, our RestClient returns `any`:

https://github.com/kumahq/kuma-gui/blob/a8c6d7fdb1f622d1f1c939f70656f4c7009bc185/src/services/kuma-api/RestClient.ts#L16-L20

So therefore it makes sense that `fromJSON*` accepts `any` i.e. we are converting the raw data "from JSON" to the type and shape we want and then `fromJSON*` returns that type once we've ensured that it is of that type.

What are you suggesting, that we just tell Typescript `any` is a `DataplaneOverview` before passing it to the methods that actually ensure that it is a DataplaneOverview shape?

I think the other thing to take into account is that the auto generated files all receive `any` so this isn't me that's made the decision to have it like this. Everybody else that uses `ts-proto` also have all these methods receive `any` so I guess its perfectly "acceptable" for them 🤷 

> Critically this prevents us from making unnecessary mistakes inside those files because you won’t be able to pass a non-conforming object.

You are kinda supposed to be able to pass in a potentially non-conforming object. The point is these methods will check it and turn it into a conforming one. Thats the whole point.

Umm I suppose to help me to understand the best question I can ask is:

> What are you suggesting, that we just tell Typescript `any` is a `DataplaneOverview` before passing it to the methods that actually ensure that it is a DataplaneOverview shape?

And if not, what is it you are suggesting?

Reply:
> What are you suggesting, that we just tell Typescript any is a DataplaneOverview before passing it to the methods that actually ensure that it is a DataplaneOverview shape?

Yes, absolutely. At this point, we can only trust that the API types we get are correct. But that’s a completely different story inside the data transformation files. We use those methods inside of those files, too. Having the type-annotated parameters ensures those methods are self-consistent.

> You are kinda supposed to be able to pass in a potentially non-conforming object. The point is these methods will check it and turn it into a conforming one. Thats the whole point.

That seems completely undesirable to me. That should not be allowed because it’s unnecessarily lax in terms of type-checking. These methods can only do their job correctly if their input is at least already conforming to the API type. If they weren’t conforming to the non-confirming type, us type-asserting that they are is an error that we wouldn’t spot because of the `any` signature.

Since inside our custom data transform files, we make use of those methods, we do actually have control over the types and can take that as an advantage.

Reply:
Ok this might help me to understand your point.

Considering the HTTP response is a string that we fetch and parse to JSON (i.e. `any`) and return from `get`, what is the difference between:

```javascript
getDataplanes(): DataplaneOverview {
  return this.get("/dataplane")
}
```

and

```javascript
fromJSON(json: any): DataplaneOverview {
  return MakeSureAnyIsActuallyADataPlaneOverview(json)
}
fromJSON(await this.get("/dataplane"))
```

i.e. when you say:

> Having the type-annotated parameters ensures those methods are self-consistent.

And you say "ensure" how are you ensuring that?

Reply:
There is no difference because this is the place at which we *fundamentally cannot have better knowledge*. However, we also use the `fromJSON*` methods inside the data transformation files and call them *which much, much better knowledge*. Having types on the parameters is essential to make those transformations rely on the type system completely _where possible_.

Reply:
> There is no difference

What about:

Given the following HTTP response:

```json
{
  "responsesSent": "2",
  "connectTime": "2021-02-17T08:33:36.442044+01:00"
}
```

and the following generated type:

```typescript
export interface Thing {
  responsesSent: number;
  connectTime: Date
}
```

Then is there a difference between:

```javascript
getThing(): Thing {
  return this.get("/thing")
}
```

and

```javascript
fromJSON(json: any): Thing {
  return MakeSureAnyIsActuallyAThing(json)
}
fromJSON(this.get("/thing"))

```

?